HW-4. Data Binding, 최근 검색 Activity 추가

onxmoreplz opened the pull request:
최근 검색어 저장 Activity 만드는 부분에서 오래걸려 늦었습니다. 죄송합니다.
최근 검색어 저장 기능 구현 부분에서 @hhi-5258 님 PR 참고했습니다.

onxmoreplz committed:
DataBinding 사용을 위한 플러그인 추가

onxmoreplz committed:
MainActivity와 activity_main 레이아웃 파일 DataBinding

onxmoreplz committed:
리싸이클러뷰 Adapter와 movie_item 레이아웃 파일 Data Binding

onxmoreplz committed:
Binding Adapter 적용

- Glide 적용을 위한 함수
- 감독이름 텍스트 수정

onxmoreplz committed:
Main 패키지 생성 후 파일 이동

onxmoreplz committed:
검색어 저장을 위한 LocalDataSource 인터페이스 정의

onxmoreplz committed:
Repository에 최근 검색어 저장,불러오기 위한 함수 정의

onxmoreplz committed:
최근 검색어 목록 Activity 및 Layout파일 생성

onxmoreplz committed:
검색어 저장 구현을 위한 SharedPreference 사용

onxmoreplz committed:
MainActivity) 넘어온 Intent Check 함수 추가

onxmoreplz committed:
MainActivity와 DataBinding 적용

onxmoreplz committed:
MainPresenter) 검색어 저장을 위한 함수 saveQuery 추가

onxmoreplz committed:
RecentSearchMovie 액티비티 생성에 따른 Contract와 Presenter 추가

onxmoreplz committed:
RecentSearchMovie 액티비티 DataBinding 적용 및 리싸이클러뷰 구현을 위한 Adapter생성

dev-donghwan commented on class14/onxmoreplz/HW2_Project/app/src/main/java/com/example/hw2_project/ViewExtensions.kt:
```diff
@@ -0,0 +1,20 @@
+
+@BindingAdapter("imageUrl")
+fun ImageView.setGlideImage(url: String) {
+    Glide.with(this.context)
```
binding adapter에서 받아오는 인자는 nullable로 받아와서 null check후에 사용해야하는걸로 알고 있습니다!

onxmoreplz replied:
그렇군요 수정했습니다!

77aeca7

dev-donghwan commented on class14/onxmoreplz/HW2_Project/app/src/main/java/com/example/hw2_project/recentSearch/RecentSearchRecyclerViewAdapter.kt:
```diff
@@ -0,0 +1,57 @@
+    inner class ViewHolder(recentMovieItemView: View) : RecyclerView.ViewHolder(recentMovieItemView){
+
+        fun bind(query: String, listener: View.OnClickListener){
+            itemView.recent_item_text_title.text = query
+            itemView.recent_item_text_title.setOnClickListener(listener)
```
여기도 databinding을 적용해 보는게 좋을거 같습니다.

onxmoreplz replied:
네! 적용했습니다!

2723baf

onxmoreplz committed:
BindingAdapter 인자 Nullable 선언 후 null check

onxmoreplz committed:
RecentSearchRecyclerViewAdapter Data binding 적용

onxmoreplz committed:
SharedPreference 로 생성되는 xml 파일 이름 변경

onxmoreplz committed:
this 중복 제거를 위한 run 함수 사용

sjjeong commented on class14/onxmoreplz/HW2_Project/app/src/main/java/com/example/hw2_project/App.kt:
```diff
@@ -0,0 +1,14 @@
+class App  : Application() {
+    companion object {
+        lateinit var prefs : SharedPreferenceUtil
+    }
```
companion object는 맨 아래로 옮겨주세요

onxmoreplz replied:
코드 위치 수정했습니다!

23c07c9

sjjeong commented on class14/onxmoreplz/HW2_Project/app/src/main/java/com/example/hw2_project/App.kt:
```diff
@@ -0,0 +1,14 @@
+
+    override fun onCreate() {
+        prefs = SharedPreferenceUtil(applicationContext)
+        super.onCreate()
```
onCreate를 먼저 호출해주세요

onxmoreplz replied:
수정했습니다.


5191460

sjjeong commented on class14/onxmoreplz/HW2_Project/app/src/main/java/com/example/hw2_project/SharedPreferenceUtil.kt:
```diff
@@ -0,0 +1,34 @@
+import android.content.SharedPreferences
+import org.json.JSONArray
+
+class SharedPreferenceUtil( context : Context) {
```
reformat code 해주세요

onxmoreplz replied:
전체적으로 reformat 했습니다!

e4a32fe

sjjeong commented on class14/onxmoreplz/HW2_Project/app/src/main/java/com/example/hw2_project/SharedPreferenceUtil.kt:
```diff
@@ -0,0 +1,34 @@
+
+        if(queryList.size > 5) {
+            queryList.removeAt(0)
+        }
```
이런 로직은 LocalDataSource쪽에서 처리 해주세요

onxmoreplz replied:
로직들 LocalDataSource에서 처리하도록 수정했습니다.

e248add

sjjeong commented on class14/onxmoreplz/HW2_Project/app/src/main/java/com/example/hw2_project/main/MainActivity.kt:
```diff
@@ -0,0 +1,75 @@
+
+    private val mainAdapter = MainRecyclerViewAdapter()
+
+    private val movieRepositoryImp = MovieRepositoryImpl()
```
변수명에서 Imp는 빼주세요

onxmoreplz replied:
변수명 수정했습니다.

9e739fb

sjjeong commented on class14/onxmoreplz/HW2_Project/app/src/main/java/com/example/hw2_project/main/MainPresenter.kt:
```diff
@@ -1,17 +1,18 @@
-        }else {
-            repositoryImpl.getMovieList(
+        } else {
+            repository.saveQuery(query)
```
save는 여기서 호출하지 마시고 repository에서 처리해주세요

onxmoreplz replied:
save기능 함수 위치 Repository로 이동했습니다.

cc569f8

sjjeong commented on class14/onxmoreplz/HW2_Project/app/src/main/java/com/example/hw2_project/main/MainRecyclerViewAdapter.kt:
```diff
@@ -0,0 +1,43 @@
+    private val arrListOfMovie :ArrayList<NaverMovieData.NaverMovie> = ArrayList()
+
+    fun updateMovieList(movies : List<NaverMovieData.NaverMovie>){
+        this.arrListOfMovie.clear()
+        this.arrListOfMovie.addAll(movies)
```
with나 run으로 묶어줄 수 있습니다

onxmoreplz replied:
with으로 묶었습니다! 감사합니다!

17f3101

onxmoreplz replied:
코드가 동작을 안해서 run함수로 바꿨습니다!

307f2ab

sjjeong commented on class14/onxmoreplz/HW2_Project/app/src/main/res/layout/activity_recent_search.xml:
```diff
@@ -0,0 +1,43 @@
+            android:text="돌아가기"
+            android:textSize="20dp"
+            android:layout_gravity="center"
+            android:onClick="@{() -> recentSearchActivity.clickBackBtn()}"/>
```
이 버튼 동작하나요?

onxmoreplz replied:
동작을 안하네요. 함수 참조도 되는데 실행이 안되는 이유가 뭔가요?

onxmoreplz replied:
동작을 안하네요...제가 놓치고 있는 부분이 뭘까요...

sjjeong replied:
recentSearchActivity가 null이에요
binding.recentSearchActivity = this 추가해주면 됩니다

onxmoreplz replied:
감사합니다...놓쳐서는 안될거를 놓치고있었네요...
수정했습니다!


0cf3366

sjjeong commented on class14/onxmoreplz/HW2_Project/app/src/main/java/com/example/hw2_project/recentSearch/RecentSearchRecyclerViewAdapter.kt:
```diff
@@ -0,0 +1,59 @@
+    inner class ViewHolder(private val recentMovieItemBinding: RecentMovieItemBinding) : RecyclerView.ViewHolder(recentMovieItemBinding.root){
+        fun bind(query: String, listener: View.OnClickListener){
+            recentMovieItemBinding.recentItemTextTitle.text = query
+            recentMovieItemBinding.recentItemTextTitle.setOnClickListener(listener)
```
리스너는 bind시점이 아니라 create시점에 추가하도록 해주세요

onxmoreplz replied:
listener 추가 위치 변경하면서 click시 호출되는 함수 부분에 영화 text 전달하기 위해 인터페이스에 함수 만들어 오버라이딩했습니다.

추가로, MainActivity에서 액티비티 호출 함수 변경했습니다.(startActivity -> startActivityForResult)

onxmoreplz committed:
Code reformat

onxmoreplz committed:
companion object 코드 위치 수정

onxmoreplz committed:
onCreate함수 먼저 호출하도록 수정

onxmoreplz committed:
query List with으로 묶기

onxmoreplz committed:
변수명 변경( Impl 제거)

onxmoreplz committed:
updateMovieList 함수) 검색목록 List run 함수로 묶기

onxmoreplz committed:
saveQuery함수 위치 변경(MainPresenter -> MovieRepository)

onxmoreplz committed:
SharedPreference 관련 로직 처리 코드 Repository로 이동

onxmoreplz committed:
RecentRecyclerViewAdapter ) listener 삭제

액티비티 호출 방식을 StartActivityForResult로 변경함에 따라 
  1) bind함수의 listener 파라미터 삭제
  2) 클릭한 영화제목 String 전달을 위한 SavedMovieClickListener 인터페이스 생성
  3) SavedMovieClickListener 의 clickSavedMovie함수 오버라이딩하여 Intent생성 후 전달

onxmoreplz committed:
RecentSearchActivity 바인딩

sjjeong reviewed (approved):
LGTM ✊ ✌️ 👋
